### PR TITLE
[lsp-pwsh] Use lsp-protocol

### PR DESCRIPTION
**Pull request by ericdallo**
https://github.com/emacs-lsp/lsp-mode/issues/1751

### CODE REVIEW

File: lsp-pwsh.el [hunk 1]
```diff
@@ -256,25 +260,24 @@ Must not nil.")
...
+(lsp-defun lsp-pwsh--apply-code-action-edits ((&pwsh:ApplyCodeActionEdits :command :arguments))
   "Handle ACTION for PowerShell.ApplyCodeActionEdits."
-  (-if-let* ((command (gethash "command" action))
-             ((&hash "StartLineNumber" "EndLineNumber"
-                     "StartColumnNumber" "EndColumnNumber" "Text")
-              (lsp-seq-first (gethash "arguments" action)))
+  (-if-let* (((&pwsh:ApplyCodeActionEditsArguments :start-line-number :end-line-number
+                                                   :start-column-number :end-column-number :text)
+              (lsp-seq-first arguments))
              (edits `[,(ht ("range" (ht ("start"
```
Comment (yyoncho) [inline lsp-pwsh.el:268]: ht usage should be fixed as well. There are 3 options:

1. Use `lsp-set-*` to update the structures inplace
2. Do the work on top of plists which has generated contractors as well
3. Me merging the lsp-protocol.el change from plist.

File: lsp-pwsh.el [hunk 2]
```diff
@@ -256,25 +260,24 @@ Must not nil.")
 (defun lsp-pwsh--extra-init-params ()
   "Return form describing parameters for language server.")
 
-(defun lsp-pwsh--apply-code-action-edits (action)
+(lsp-defun lsp-pwsh--apply-code-action-edits ((&pwsh:ApplyCodeActionEdits :command :arguments))
```
Comment (kiennq) [inline lsp-pwsh.el:263]: This should be `&Command`
https://github.com/microsoft/language-server-protocol/blob/gh-pages/_specifications/specification-3-16.md#-command-

File: lsp-pwsh.el [hunk 3]
```diff
@@ -256,25 +260,24 @@ Must not nil.")
...
-                           ("newText" Text))]))
+                                         (ht ("line" (- end-line-number 1))
+                                             ("character" (- end-column-number 1))))))
+                           ("newText" text))]))
       (lsp--apply-text-edits edits)
-    (lsp-send-execute-command command (gethash "arguments" action))))
+    (lsp-send-execute-command command arguments)))
 
-(defun lsp-pwsh--show-code-action-document (action)
+(lsp-defun lsp-pwsh--show-code-action-document ((&pwsh:ShowCodeActionDocumentation :arguments))
```
Comment (kiennq) [inline lsp-pwsh.el:278]:
```suggestion
(lsp-defun lsp-pwsh--show-code-action-document ((&Command :arguments))
```

File: lsp-pwsh.el (hunk 2 above)
Comment (ericdallo) [reply lsp-pwsh.el:263]: Thanks for noticing that

File: lsp-protocol.el [hunk 4]
```diff
@@ -153,16 +153,20 @@ See `-let' for a description of the destructuring mechanism."
 
 ;; manually defined interfaces
 (lsp-interface (RpcRequestResponse (:params :id :method) nil))
+(lsp-interface (Command (:title :command) (:arguments)))
```
Comment (kiennq) [inline lsp-protocol.el:156]: This has already being defined
```suggestion
```

File: lsp-protocol.el [hunk 5]
```diff
@@ -153,16 +153,20 @@ See `-let' for a description of the destructuring mechanism."
...
-(lsp-interface (eslint:OpenDoc (:url) nil))
+(lsp-interface (eslint:StatusNotification (:state) nil)
+               (eslint:WorkspaceConfigurationReqHandler (:items) nil)
+               (eslint:WorkspaceConfiguration (:scopeUri) nil)
+               (eslint:OpenDoc (:url) nil))
 
 (lsp-interface (haxe:ProcessStartNotification (:title) nil))
 
+(lsp-interface (pwsh:ApplyCodeActionEdits (:StartLineNumber :EndLineNumber :StartColumnNumber :EndColumnNumber :Text) nil)
+               (pwsh:ApplyCodeActionEditsRequest (:range :newText) nil))
```
Comment (kiennq) [inline lsp-protocol.el:168]: What's this used for?

File: lsp-protocol.el [hunk 6]
```diff
@@ -153,16 +153,20 @@ See `-let' for a description of the destructuring mechanism."
...
-(lsp-interface (eslint:WorkspaceConfiguration (:scopeUri) nil))
-(lsp-interface (eslint:OpenDoc (:url) nil))
+(lsp-interface (eslint:StatusNotification (:state) nil)
+               (eslint:WorkspaceConfigurationReqHandler (:items) nil)
+               (eslint:WorkspaceConfiguration (:scopeUri) nil)
+               (eslint:OpenDoc (:url) nil))
 
 (lsp-interface (haxe:ProcessStartNotification (:title) nil))
 
+(lsp-interface (pwsh:ApplyCodeActionEdits (:StartLineNumber :EndLineNumber :StartColumnNumber :EndColumnNumber :Text) nil)
```
Comment (kiennq) [inline lsp-protocol.el:167]: This should be `pwsh:ScopeRegion`.
https://github.com/PowerShell/PowerShellEditorServices/blob/3fcd0c651d50ec629b90c186c766af1872f81459/src/PowerShellEditorServices/Services/TextDocument/ScriptRegion.cs

File: lsp-protocol.el (hunk 5 above)
Comment (ericdallo) [reply lsp-protocol.el:168]: It'll be used with `lsp-make-*` when the `plists` branch got merged. I'll remove until the merge.

File: lsp-protocol.el (hunk 6 above)
Comment (ericdallo) [reply lsp-protocol.el:167]: `ScriptRegion` you mean, right?
Thanks, I found the repo, but could not find these definitions.

File: lsp-protocol.el (hunk 4 above)
Comment (ericdallo) [reply lsp-protocol.el]: Ops, forgot about that

File: lsp-pwsh.el [hunk 7]
```diff
@@ -164,7 +168,7 @@ For more information about the brace styles please refer to https://github.com/P
   :package-version '(lsp-mode . "6.2"))
 
 (defcustom lsp-pwsh-developer-editor-services-log-level "Normal"
-  "Sets the logging verbosity level for the PowerShell Editor Services host executable.
+  "Sets the logging verbosity level for the PWSH Editor Services host exe.
```
Comment (kiennq) [inline lsp-pwsh.el:171]:
```suggestion
  "Sets the logging verbosity level for the PowerShell Editor Services host exe.
```

Comment (kiennq) [reply lsp-pwsh.el:171]: Also, this sentence is from VsCode plugin customization, I think we should keep this as it.

File: lsp-protocol.el (hunk 5 above)
Comment (kiennq) [reply lsp-protocol.el]: In case you haven't notice, it's `TextEdit` type

**Comment by ericdallo** [comment]
Depends on https://github.com/emacs-lsp/lsp-mode/pull/1765

**Comment by kiennq** [comment]
@ericdallo @yyoncho Can we make this PR against emacs-lsp/lsp-mode#plists branch instead?
That branch definitely need each language client support before it can be merged.

**Comment by ericdallo** [comment]
Agree, I can make another PR pointing to lists branch

**Comment by ericdallo** [comment]
Done: https://github.com/emacs-lsp/lsp-mode/pull/1780
Should we close this?

**Comment by kiennq** [comment]
I think so

**Pull request closed without merging**